### ISSUE TITLE
Tracking checkpoints or blocks

### ISSUE THREAD

**Issue by perfectmak** [snapshot-labs/sx-api#36]
Currently, there is a table `checkpoints` that is used to keep track of blocks already visited.

It is created with the following structure:

```sql
CREATE TABLE checkpoint (
	number BIGINT NOT NULL,
	PRIMARY KEY (number)
);
```

When `Checkpoint.start()` is invoked, a new record (with `number` set to 0) is inserted into the `checkpoints` table.

After each block is fetched, the `number` in the record is updated to the blocks value.

```sql
UPDATE checkpoint SET number = <number>
```

 The algorithm works like this:

- Fetch starting block
    - If there exists a `checkpoints.number` value in the database and this value is greater than all the `start` block in the config.sources list, then this value + 1 is used.
    - Else, we use the lowest `start` block specified in the config.sources list or default to 0 is none is specified.
- Parse block for config events
- Call relevant writer handler functions
- Update `checkpoints.number`  with last block processed.
- Repeat from step 2 with next block number

## Why do we need better block tracking:

To track blocks that contain relevant snapshot events, so that Checkpoint can use it as a skip list when scanning blocks. 

Also, if relevant blocks are tracked in the database, it will be easy to create snapshots (no pun intended 😅) of the table for backup and recovery (or just transferring between nodes).

## Proposed Solution

The proposed solution is to keep track of all blocks where we find a relevant

Change the checkpoint schema to something like:

```sql
CREATE TABLE _checkpoints (
	block_number BIGINT NOT NULL,
	contract_address VARCHAR(65), -- maybe store this as bytes to save space
	PRIMARY KEY (block_number, contract_address)
);

-- we definitely can normalize checkpoint_block_events to reduce duplications

CREATE TABLE _metadata (
	id VARCHAR(20) NOT NULL,
	value VARCHAR(128) NOT NULL,
	PRIMARY KEY (id)
);
```

> NOTE: Adopting the `_` prefix to signify core tables and reduce the chances of users entities conflicting with core tables names.
> 

The `_checkpoints` table will be used to record blocks where we find events related to a contract.

The `_metadata` is a generic table meant to store key-value pairs of information that will be relevant during the lifecycle of Checkpoint’s operations. I’ll explain more on what will be stored here in the algorithms below.

To make things more efficient, the algorithm for fetching and processing blocks will need to change. The updated algorithm should run two asynchronous functions when the `Checkpoint.start()` function is invoked.

### 1. BlockIndexer

The BlockIndexer focuses on fetching blocks, confirming it contains the relevant contracts (based on the config the user provides) and recording this information in the `_checkpoints` table.

The block indexer when started will do the following steps:

- Fetch starting block.
    - It checks if there is a row in the `_metadata` table with id `last_indexed_block` and uses that value + 1.
    - If not, then it starts from block 0.
- Parse block for config events
    - Fetch the block from the network
    - Validate and confirm if relevant events exists in block
    - If found, insert a new row with contract address, block number into the `_checkpoints` table. NOTE: If it is a deploy function use the `::deploy` special value as the `event` columns value.
- Update `_metadata`'s `last_indexed_block` value  to the last block processed.
- Repeat from step 2 with next block number

> NOTE: An advantage of tracking a `last_indexed_block` value is to give users the flexibility of changing the start block and force-reindexing
> 

### 2. WriteInvoker

The WriteInvoker will be polling the `_checkpoints` table for new blocks and invoking the respective `writer[fn]` with the data for the new blocks.

Because only `block_numbers` exist in the database, this means the WriteInvoker will still need to fetch the full blocks from the network. This could lead to some inconsistency in cases where the blocks are not final, but for now it shouldn’t be too much of a problem as re-processing should be straight forward.

After a writer fn has been successfully invoked, `_metadata` row with `id` `last_invoked_block` will be set to the last block, before moving on to the next block.

## Ensuring Consistency

It is possible that a user stops the Checkpoint process and restarts it with an updated config containing new contracts and events. For now, the easiest solution will be call the `Checkpoint.reset()` method before calling the `Checkpoint.start()` method, so everything starts on a clean state.

A future improvement would be that before `Checkpoint.start()` starts the indexing, it computes a hash based off the `contract` and `event` names in the config and stores it as a row in `_metadata`. So on subsequent starts, it compares if the hash in the database is the same as the hash of the new configs. If they hashes are not the same it can throw an error.

**Comment by bonustrack** [comment]
Here is some quick feedback:
- For the table prefix I would just use `_` instead of `checkpoint_`, having this start with _ make it more obvious it's part of the core, it would appear first on MySQL admin UIs. 
- For `checkpoint_block_events` the primary key won't work cuz there can be multiple events with the same name / contract / block. I would rather not store event name but only store if a contract was seen on a block or not. This would reduce the number of items in the db and make it optimal to query checkpoints. 
- I would not use a name "block_events" cuz we would later support storing data from tx (and not just events), this is something we already do in some way with `deployFn`. I would just call it `checkpoints` this is what is it actually.
- For BlockIndexer and WriteInvoker what you trying to address with this change? The main point we should optimise is speed, we need to be able to sync asap but also get live data asap. Currently the slowest part is not write but loading blocks, from my understanding with BlockIndexer and WriteInvoker you would load a single block twice, and the process will be divided in 2 which doesn't seem optimal.

**Comment by perfectmak** [comment]
> For checkpoint_block_events the primary key won't work cuz there can be multiple events with the same name / contract / block. I would rather not store event name but only store if a contract was seen on a block or not. This would reduce the number of items in the db and make it optimal to query checkpoints.

You are right, I wrongly assumed single events per contract per block. I’ll update accordingly.

> For BlockIndexer and WriteInvoker what you trying to address with this change? The main point we should optimise is speed, we need to be able to sync asap but also get live data asap. Currently the slowest part is not write but loading blocks, from my understanding with BlockIndexer and WriteInvoker you would load a single block twice, and the process will be divided in 2 which doesn't seem optimal.

The reason I propose separating is to make the indexing more robust and independent from writers. We don’t have to wait for the writer to complete each time before moving to the next block and building up the checkpoints. 

True, loading the blocks twice is not optimal. But ultimately we will still need to address the block fetching speed later maybe by introducing a cache at one point, so I didn’t think it was going to be too much of an issue now. 

We can definitely merge the two steps together like it currently is and plan more for this optimization later.

**Comment by bonustrack** [comment]
The fetching of block twice can be annoying if it create delay when the chain is stream to latest block and we look for new blocks. When you see your tx got accepted on the UI you expect getting the information stored on the API directly when you refresh the page, anything we can to remove latency here is great, that's why I'm not too much into make it with 2 processes. We could separate this into 2 classes but IMO it should be made into a single process.

**Comment by perfectmak** [comment]
The core functionality of this proposal was implemented in #39. We can open new issues for subsequent improvements.
